resnet/main.py

```diff
@@ -0,0 +1,238 @@
+            crop_size=config.train_dataset.crop_size,
+            num_workers=8,
+            pin_memory=True,
+            persistent_workers=True)
+
+    print('Built train dataloader\n')
+
+    # Validation dataset
+    print('Building evaluation dataloader')
+    if config.eval_dataset.is_streaming:
+        eval_dataspec = build_streaming_imagenet_dataspec(
+            remote=config.eval_dataset.path,
+            local=config.eval_dataset.local,
+            batch_size=eval_batch_size,
+            is_train=False,
+            drop_last=False,
+            shuffle=False,
+            resize_size=config.eval_dataset.resize_size,
+            crop_size=config.eval_dataset.crop_size,
+            num_workers=8,
+            pin_memory=True,
+            persistent_workers=True)
+    else:
+        eval_dataspec = build_imagenet_dataspec(
+            datadir=config.eval_dataset.path,
+            batch_size=eval_batch_size,
+            is_train=False,
+            drop_last=False,
+            shuffle=False,
+            resize_size=config.eval_dataset.resize_size,
+            crop_size=config.eval_dataset.crop_size,
+            num_workers=8,
+            pin_memory=True,
+            persistent_workers=True)
+    print('Built evaluation dataloader\n')
+
+    # Instantiate torchvision ResNet model
+    print('Building Composer model')
+    composer_model = build_composer_resnet(model_name=config.model.name,
+                                           loss_name=config.model.loss_name,
+                                           num_classes=config.model.num_classes)
+    print('Built Composer model\n')
+
+    # Optimizer
+    print('Building optimizer and learning rate scheduler')
+    optimizer = DecoupledSGDW(composer_model.parameters(),
+                              lr=config.optimizer.lr,
+                              momentum=config.optimizer.momentum,
+                              weight_decay=config.optimizer.weight_decay)
+
+    # Learning rate scheduler: LR warmup for 8 epochs, then cosine decay for the rest of training
+    lr_scheduler = CosineAnnealingWithWarmupScheduler(
+        t_warmup=config.scheduler.t_warmup, alpha_f=config.scheduler.alpha_f)
+    print('Built optimizer and learning rate scheduler\n')
+
+    # Callbacks for logging
+    print('Building Speed, LR, and Memory monitoring callbacks')
+    speed_monitor = SpeedMonitor(window_size=50) # Measures throughput as samples/sec and tracks total training time
+    lr_monitor = LRMonitor() # Logs the learning rate
+    memory_monitor = MemoryMonitor() # Logs memory utilization
+
+    # Callback for checkpointing
+    print('Built Speed, LR, and Memory monitoring callbacks\n')
+
+    # Recipes for training ResNet architectures on ImageNet in order of increasing training time and accuracy
+    # To learn about individual methods, check out "Methods Overview" in our documentation: https://docs.mosaicml.com/
+    print('Building algorithm recipes')
```

Comment:
Don't we need to overwrite parts of the config based on the selected recipe? E.g., the recipe's train and eval crop sizes don't seem to be used.

Probably has to happen right when we load the config, at the bottom of the file.

I think [omegaconf accepts nested "." attributes in key names](https://omegaconf.readthedocs.io/en/2.2_branch/usage.html#omegaconf-update), so the overwriting should be pretty simple.

Reply:
I dropped the ball on this one. I was testing doing it at the bottom, but that would make it hard to test recipes in the `test_trainer` script. Any objections to adding it to the top of `main()`?

Reply:
I added the recipe update at the top of `main()` and added tests for the recipe names in `test_trainer`. I'm not sure if changes the parameters under the user is the right interface, but I think we should iterate on it instead of delay the PR